:sparkles: Apply PodSecurityStandard in Secure Cluster Class (#6390)
kubernetes-sigs/cluster-api: docs/book/src/security-guidelines/pod-security-standards.md

# Pod Security Standards

> Pod Security Admission allows defining [Pod Security Standards] for creation of pods at the cluster level.

Reviewer: ```suggestion
Pod Security Admission allows applying [Pod Security Standards] during creation of pods at the cluster level.
```

```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: ClusterClass
spec:
  variables:
  - name: podSecurityStandard
    required: false
    schema:
      openAPIV3Schema:
        type: object
        properties: 
          enabled: 
            type: boolean
            default: true
            description: "enabled enables the patches to enable Pod Security Standard via AdmissionConfiguration."
          enforce:
            type: string
            default: "baseline"
            description: "enforce sets the level for the enforce PodSecurityConfiguration mode. One of privileged, baseline, restricted."
            pattern: "privileged|baseline|restricted"
          audit:
            type: string
            default: "restricted"
            description: "audit sets the level for the audit PodSecurityConfiguration mode. One of privileged, baseline, restricted."
            pattern: "privileged|baseline|restricted"
          warn:
            type: string
            default: "restricted"
            description: "warn sets the level for the warn PodSecurityConfiguration mode. One of privileged, baseline, restricted."
            pattern: "privileged|baseline|restricted"
  ...
```

> 

Reviewer: Maybe worth adding a couple of notes:
- The version field in Pod Security Admission Config defaults to `latest` 
- The `kube-system` namespace is exempt from Pod Security Standards enforcement, because it runs control-plane pods that need higher privileges

        valueFrom:
          template: |
            - content: |
                apiVersion: apiserver.config.k8s.io/v1
                kind: AdmissionConfiguration
                plugins:
                - name: PodSecurity
                  configuration:
                    apiVersion: pod-security.admission.config.k8s.io/v1beta1
                    kind: PodSecurityConfiguration
                    defaults:
                      enforce: "{{ .podSecurity.enforce }}"
                      enforce-version: "latest"
                      audit: "{{ .podSecurity.audit }}"
                      audit-version: "latest"
                      warn: "{{ .podSecurity.warn }}"
                      warn-version: "latest"
                    exemptions:
                      usernames: []
                      runtimeClasses: []
                      namespaces: [kube-system]
              path: /etc/kubernetes/kube-apiserver-admission-pss.yaml
    enabledIf: "{{ .podSecurityStandard.enabled }}"
...
```

{{#/tab }}
{{#/tabs }}

[Pod Security Standards]: https://kubernetes.io/docs/concepts/security/pod-security-standards

### Create a secure Cluster using the ClusterClass

> After adding the variables and patches the Pod Security Standard would be added by default.

Reviewer: ```suggestion
After adding the variables and patches the Pod Security Standards would be applied by default.
```

                      audit-version: "latest"
                      warn: "{{ .podSecurity.warn }}"
                      warn-version: "latest"
                    exemptions:
                      usernames: []
                      runtimeClasses: []
                      namespaces: [kube-system]
              path: /etc/kubernetes/kube-apiserver-admission-pss.yaml
    enabledIf: "{{ .podSecurityStandard.enabled }}"
...
```

{{#/tab }}
{{#/tabs }}

[Pod Security Standards]: https://kubernetes.io/docs/concepts/security/pod-security-standards

### Create a secure Cluster using the ClusterClass

After adding the variables and patdocs/book/src/tasks/experimental-features/cluster-class/write-clusterclass.mdches the Pod Security Standards would be applied by default.
It is also possible to disable this patch or configure different levels for the configuration 
using variables.

```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: Cluster
metadata:
  name: "my-cluster"
spec:
  ...
  topology:
    ...
    class: my-secure-cluster-class
    variables:
    - name: podSecurityStandard
      value: 
        enabled: true
        enforce: "restricted"
        audit: "restricted"

>         warn: "restricted"

Reviewer 2: nit: Should we drop this given that `restricted` is the default
